### ISSUE TITLE
Neo Geo CD support missing

### ISSUE THREAD

**Issue by Oggom** [libretro/libretro-fba#4]
The standalone release of FBA supports playing Neo Geo CD games, which hasn't been implemented in the libretro port yet.
https://github.com/libretro/fba-libretro/issues/28

**Comment by Zombie-Ryushu** [comment]
I'd like to bump this issue.

**Comment by duskgao** [comment]
@Oggom   Anyone just port NJEMU to libretro!?

https://github.com/phoe-nix/NJEMU

 :D

**Comment by Oggom** [comment]
@duskgao Hooking up NGCD emulation would probably be easier than porting over an entirely different emulator to libretro.
However, I think there actually have been talks about porting over NJEMU since FBA is too slow for the PSP port of RA, but I'm not sure if anything ever came out of it.

**Comment by fr500** [comment]
Actually I added njemu-cps2 to the buildbot last night

**Comment by duskgao** [comment]
@fr500  any changs about njemu-ngcd ??

**Comment by fr500** [comment]
njemu is only for PSP.. and I dunno, the main dev for that core is @aliaspider but I haven't seen him in a while

**Comment by duskgao** [comment]
@fr500 @Oggom thx for all your reply...It is really abnormal that there is no ngcd emulator for android in such a long time  :(

**Comment by Redemp** [comment]
Any more news on Neo Geo CD support or has it been forgotten?

**Comment by barbudreadmon** [comment]
Not forgotten but beyond my knowledge, i'm kinda good at updating things and tracking/debugging small bugs, but lacking when it comes to adding new features.

I think lines 324-336 of libretro.cpp are related to this feature, not even sure.

Perhaps someone like @AkimanBengus could help.

**Comment by AkimanBengus** [comment]
This is a feature which I would also like to have but I'm really not able to include it.
I'm a newbie in emulation programming :-(

**Comment by barbudreadmon** [comment]
Is there any trick about this feature on standalone ? I wanted to figure out how it works on standalone, but i fail at launching any of my neogeo cd iso.

**Comment by Redemp** [comment]
I have seen that you get alot of help from the fba developers on the forums. Maybe they can help

**Comment by shad00m** [comment]
a sad +1 :(

**Comment by Darknior** [comment]
Some news about this upgrade to add Neogeo CD ?
There are only 4 cool games on it, but i really dream to play Samurai Shodown RPG 🗡

**Comment by barbudreadmon** [comment]
4 games ? Which ones ? Right now that seems like a lot of work for 1 untranslated japanese rpg, and i don't see any other game unplayable without neogeo cd support.

**Comment by Zombie-Ryushu** [comment]
There are more games than just four. Most AES, and MVS Neo-Geo games have a Neo-CD Counterpart.  These games have expanded music, or different music than the Cart versions. Such as PulStar. Support for Neo-CD means better audio for the Metal Slug games in alot of cases that are typically found on Collections for the Wii.

**Comment by barbudreadmon** [comment]
There are only 8 exclusive titles to neogeo cd : 
- The King of Fighters '96 Collection which is not a game
- ADK World which is a promotional cd
- Neogeo CD Special which is a promotional cd
- Samurai Showdown RPG
- Crossed Swords 2 which is available on mvs
- Mahjong Final Romance 2 which is available on mvs
- Zintrick which is available on mvs
- Ironclad which is available on mvs

So only Samurai Showdown RPG is an exclusive game to the neogeo cd.
For all the non-exclusive games, i don't think many people would give up having no loading time just to have different audio, that's why fixing neogeo cd support looks like really low priority to me, so don't count me in as long as there are other issues (and there are a few : some games crash on 64bits, i really want to fix the tree-loop-vectorize issue with optimization, and there is still some work to do on inputs)

**Comment by Darknior** [comment]
@barbudreadmon you are real.
The only cool games to play on NeoGeo CD are 

- Samurai Showdown RPG
- Crossed Swords 2 which is available on mvs
- Mahjong Final Romance 2 which is available on mvs
- Zintrick which is available on mvs
- Ironclad which is available on mvs

But it's really hard to found them in MVS :(
If you know the place to found them tell it to us ?
Impossible to found the Crossed Sword 2 MVS rom for me !!!

For Samurai RPG yes it is a Japan game, but for me a full french Translation is available, and an US one is coming too ...
We can found the game on PSX and Saturn too, but nothing was translated there :(

**Comment by Zombie-Ryushu** [comment]
Let me just weigh in here. I think we should have Neo CD Support simply because its the right thing to do.

**Comment by barbudreadmon** [comment]
@Darknior : Easy, just google "fba 0.2.97.39 crossed sword 2". Also make sure you don't use unibios with those 4 games, i'm almost sure they don't work with unibios.

@Zombie-Ryushu : Sure, for the sake of completion, feel free to work on it, right now i don't have tens of hours to spend on trying to fix a feature which will only allow me to play one more game, especially since it is not really considered a great game.

**Comment by vjarvens** [comment]
No-one mentioned Viewpoint yet :o 
https://www.youtube.com/watch?v=OfHaqaNQATg
One of my favourite games ever. Amazing soundtrack!
This was the only release with cd-quality soundtrack, please don´t mention ps1 conversion, worst ever.

Pulstar, amazing soundtrack, only on neo geo cd:
https://youtu.be/5e1Rkrxkd10?t=120

Windjammers neo geo cd also rare strong release. (only ps4 & ps vita re-releases)
https://www.youtube.com/watch?v=CPLo36bEIWQ

Twinkle star sprites (with cd audio) (also one of the alltime faveys) you can find version on saturn & dreamcast.

When you play with better soundsystem, music gives alot to immersion!

// Ville

**Comment by darksaviorx** [comment]
I was looking to play kof94 with the great cd audio in retropie, and stumbled onto this thread. Too bad it's low priority. The early games like kof94 and 95 don't have load times in between rounds so most people won't mind.

**Comment by barbudreadmon** [comment]
neogeo cd support was added, there could be issues with non-linux platform, several things to know : 
- you need a copy of neocdz.zip and neogeo.zip in your libretro system directory
- you need to add `--subsystem neocd` to the command line
- you can't load MODE1/2352 tracks (i'll try to fix this later)

**Comment by hizzlekizzle** [comment]
@barbudreadmon whoa, good stuff! I never expected this to get added :O

**Comment by thatman84** [comment]
Sounds cool,

Is there anymore detail on how to activate the subsystem when launching directly with RetroArch. Is something needed for playlist support . The referenve to it is a little above basic users heads imo.

**Comment by barbudreadmon** [comment]
@thatman84 Not sure how it works with playlist (i'm a EmulationStation user, so i never used this feature). Perhaps @fr500 can help with this.

**Comment by fr500** [comment]
There is no subsystem support for playlists, the playlist format needs an overhaul to allow extra fields for that, maybe they should use a format that allows adding fields so the format doesn't break evey time a new value needs to be added.

Also some stuff needs to be added to https://github.com/libretro/RetroArch/blob/ad79fbe4a9d561dc7d77437cbe0ef5e202950c93/dynamic.c so RetroArch itself can load content from subsystems before loading a dummy rom, as far as I can tell on "Load Core" only this function is called:

`void retro_get_system_info(struct retro_system_info *info)`

And subsystems are set in 
`void retro_set_environment(retro_environment_t cb)`

This is something only @Autechre64 or maister can probably fix

**Comment by pepodmc3** [comment]
can you add an option to multiply cd read speed like Raine Emulator?

**Comment by barbudreadmon** [comment]
@pepodmc3 Nope, i already looked into that, you can use retroarch's fast forward though.

**Comment by thorias** [comment]
@barbudreadmon any possibility to contact the NeoRaine author to help/assist in implementing it (multiply cd read speed)? In case: https://github.com/zelurker/raine
As I'm not sure about the retroarch fast forward efficiency is (regarding not having the game crashing if being used).

**Comment by thorias** [comment]
@fr500 with neogeo cd emulation?

**Comment by fr500** [comment]
no, there is no reason fast forward would make that crash though...
it just removes vsync

**Comment by thorias** [comment]
@fr500 ok fair enough then! Thank you for the details.

**Comment by jhdgghost** [comment]
One question is it CDDA music working? I try different CUE ISO/WAV and they work perfectly but no music. Only to confirm I am not missing an option or something.

**Comment by darksaviorx** [comment]
@jhdgghost I couldn't get this working with cd audio. I went back to lr-mess. That supports cd audio and chd. If using a pi, it might take over a day to compile. I got a copy from someone that waited that long.

**Comment by Boviligg** [comment]
Any news on this running in windows? I can get into the bios but games refuse to start.

**Comment by barbudreadmon** [comment]
@Boviligg I'm not aware of any issue in windows, most likely your iso is just not compatible with the emulator. Try it in standalone version and if it does run, i'll take a look at your issue.

**Comment by Boviligg** [comment]
It works in standalone complete with music. 
I just managed to boot the game in the retroarch core by first creating a memory card in the bios. Now the game boots but no music - which is strange coz the music is fine in standalone.
Thanks for your reply btw.

**Comment by twinaphex** [comment]
You get no redbook audio probably because it hasn't been implemented yet -

https://github.com/libretro/fbalpha/blob/master/src/burner/libretro/cd_emu.cpp#L82

It should be possible to either use rwave from libretro-common or some other wave decoding code elsewhere. Try seeing how the other FBA ports do it.

**Comment by Boviligg** [comment]
Cheers mate

**Comment by shinra358** [comment]
> you need a copy of neocdz.zip and neogeo.zip in your libretro system directory

So if you have the option on to put bios into core folders, it would be system/fbalpha/ correct? Or would the fbalpha folder be named something else?


> you can't load MODE1/2352 tracks (i'll try to fix this later)

Does this mean the roms with all the music tracks on the outside of the iso? Because that's all I have and no neogeo cd games are loading.

**Comment by barbudreadmon** [comment]
> So if you have the option on to put bios into core folders, it would be system/fbalpha/ correct? Or would the fbalpha folder be named something else?

It's the system/fba/ folder

> Does this mean the roms with all the music tracks on the outside of the iso?

Generally, yes.

**Comment by blackman91** [comment]
Does roms with an .iso .cue and .mp3 files work? I tried Crossed Swords 2 but the music is not playing at all.

**Comment by blackman91** [comment]
I tried a bunch of different roms but it seems the core can't play the .mp3 tracks or .wav tracks, I even tried an iso with the music files merged and the music also didn't work, has anyone got the music to work?

**Comment by blackman91** [comment]
Tested standalone and it can read the .mp3 tracks just fine, any chance this can be added to the libretro port? mp3 files are a lot smaller than  wav.
Also this core works really great on Android too, performance is great much better than MAME, the same steps work on Android:
https://forums.libretro.com/t/guide-play-neo-geo-cd-iso-games-using-fbalpha/17735

**Comment by duskgao** [comment]
anyone tested Samurai Spirits RPG on android with FBA core? 

when I load ISO or CUE,I can run the BIOS well,but after the NGCD loading screen, it just go black, nothing happen.

test with both FBA and FBA2012.

this is my cue:

FILE "Samurai Spirits RPG - Track 01.iso" BINARY
  TRACK 01 MODE1/2048
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 02.mp3" MP3
  TRACK 02 AUDIO
    PREGAP 00:03:00
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 03.mp3" MP3
  TRACK 03 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 04.mp3" MP3
  TRACK 04 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 05.mp3" MP3
  TRACK 05 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 06.mp3" MP3
  TRACK 06 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 07.mp3" MP3
  TRACK 07 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 08.mp3" MP3
  TRACK 08 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 09.mp3" MP3
  TRACK 09 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 10.mp3" MP3
  TRACK 10 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 11.mp3" MP3
  TRACK 11 AUDIO
    PREGAP 00:00:00
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 12.mp3" MP3
  TRACK 12 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 13.mp3" MP3
  TRACK 13 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 14.mp3" MP3
  TRACK 14 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 15.mp3" MP3
  TRACK 15 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 16.mp3" MP3
  TRACK 16 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 17.mp3" MP3
  TRACK 17 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 18.mp3" MP3
  TRACK 18 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 19.mp3" MP3
  TRACK 19 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 20.mp3" MP3
  TRACK 20 AUDIO
    PREGAP 00:00:00
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 21.mp3" MP3
  TRACK 21 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 22.mp3" MP3
  TRACK 22 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 23.mp3" MP3
  TRACK 23 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 24.mp3" MP3
  TRACK 24 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 25.mp3" MP3
  TRACK 25 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 26.mp3" MP3
  TRACK 26 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 27.mp3" MP3
  TRACK 27 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 28.mp3" MP3
  TRACK 28 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 29.mp3" MP3
  TRACK 29 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 30.mp3" MP3
  TRACK 30 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 31.mp3" MP3
  TRACK 31 AUDIO
    INDEX 01 00:00:00
FILE "Samurai Spirits RPG - Track 32.mp3" MP3
  TRACK 32 AUDIO
    INDEX 01 00:00:00

**Comment by hizzlekizzle** [comment]
We're getting into generalized support questions now, which github issue trackers aren't really intended for. I suggest we close this issue, since the support was added successfully, and move the support questions to the forum.

**Comment by blackman91** [comment]
But the music is not working, you think we should open another issue?

**Comment by barbudreadmon** [comment]
@twinaphex @fr500 @hizzlekizzle So i took a look at this rwave thingy, but since i didn't find an example of implementation, it didn't end well (segfaulting), i have commented the code causing the crash in https://github.com/libretro/fbalpha/blob/f547bf67dd6d043add2c2c17165f64ccd27df4ba/src/burner/libretro/cd_emu.cpp#L87-L127

As i stated several times, i have no interest in neocd (i prefer "no loading time" and "improved graphics" over "improved sound", honestly sound in neogeo titles is not that great...). It would be nice to fix it for completion purpose, but honestly if no one can help, it'll probably stay that way.

**Comment by fr500** [comment]
ah so standalone doesn't have neogeo cd audio on linux I guess

**Comment by barbudreadmon** [comment]
Finally added support for the audio tracks in https://github.com/libretro/fbalpha
Closing this issue, feel free to open a new one in the new repo if something is not working properly.

On a sidenote, idk what it's worth, but i heard there is a neocd libretro core at https://github.com/fabrice-martinez/neocd_libretro

**Comment by Darknior** [comment]
Thanks for the NeogeoCD core information ... i forgot this information :p 
I will try it, i don't know if it was already added to Retropie or not ?

**Comment by bslenul** [comment]
Awesome news, thanks for the update! Works like a charm now 👍 

I also tried the new core (on a W10 PC), works like a charm too + it seems to skip the loadings (edit: yup, there's a core option to skip loadings!) <3

**Comment by blackman91** [comment]
Updated fba core no year using online updater but the music of crossed swords 2 is still not playing, I have the .cue .iso and .mp3 files in the same folder, does it work for you @bslenul ? Tested on both windows and Android.

**Comment by bslenul** [comment]
@blackman91 I just tried Crossed Swords 2, musics are playing for me, yes. I have the tracks in .wav format however, not .mp3!

**edit:** btw, tested on W10 and on a Linux VM, works fine on both.

**Comment by blackman91** [comment]
You are right just tried a rom with .wav tracks and it worked, so no support for mp3 tracks?

**Comment by barbudreadmon** [comment]
@blackman91 Not sure why mp3 is not working, i actually implemented it, might be missing something trivial, however is that important ? You trade graphic quality and loading time for audio quality (which is already a bad deal imho, i never heard anything great in a neogeo game), then you want to cut off this audio quality ?

Well, anyway, as i already said, you can open a new issue in the new repo if you think something is not working properly.
